cmd/server/main.go

```diff
@@ -198,50 +195,51 @@ func main() {
 	if viper.IsSet("grpc_conn_max_seconds") {
 		grpcKeepAliveFor = time.Second * time.Duration(viper.GetInt("grpc_conn_max_seconds"))
 	}
-	clientBuilder := authorizer.NewClientBuilder(parseClientConfig())
 
 	authorizer, err := authorizer.NewManager(
-		clientBuilder,
+		authorizer.NewClientBuilder(parseClientConfig()),
 		createSystemCache(),
 		createBackendConfig(),
 	)
 	if err != nil {
-		istiolog.Errorf("Unable to create authorizer: %v", err)
-		os.Exit(1)
+		log.Fatalf("Unable to create authorizer: %v", err)
 	}
 
 	adapterConfig := threescale.NewAdapterConfig(authorizer, grpcKeepAliveFor)
 
 	s, err := threescale.NewThreescale(addr, adapterConfig)
-
 	if err != nil {
-		istiolog.Errorf("Unable to start sever: %v", err)
-		os.Exit(1)
-	}
-
-	if version == "" {
-		version = "undefined"
+		log.Fatalf("Unable to start sever: %v", err)
 	}
-	istiolog.Infof("Started server version %s", version)
 
 	shutdown := make(chan error, 1)
 	go func() {
+		if version == "" {
+			version = "undefined"
+		}
+		log.Infof("Starting server version %s", version)
 		s.Run(shutdown)
 	}()
 
 	sigC := make(chan os.Signal, 1)
-	signal.Notify(sigC, syscall.SIGTERM)
+	signal.Notify(sigC, syscall.SIGTERM, syscall.SIGINT)
 
-	go func() {
-		_ = <-sigC
-		fmt.Println("SIGTERM received. Attempting graceful shutdown")
-		err := s.Close()
-		if err != nil {
-			fmt.Println("Error calling graceful shutdown")
-			os.Exit(1)
-		}
-		return
-	}()
+	for {
+		select {
+		case sig := <-sigC:
+			log.Infof("\n%s received. Attempting graceful shutdown\n", sig.String())
+			err := s.Close()
+
+			if err != nil {
+				log.Fatalf("Error calling graceful shutdown")
+			}
 
-	_ = <-shutdown
+		case err = <-shutdown:
+			if err != nil {
+				log.Fatalf("gRPC server has shut down: err %v", err)
+			}
+
+			log.Info("gRPC server has shut down gracefully")
```

Comment:
Shouldn't this live in an `else` clause? Otherwise `if err != nil` you'd get a fatal notification about an error while shutting down followed by "hey, no worries, it was all perfectly right!".

Reply:
log.Fatal calls into os.Exit(1)

Reply:
That is... unexpected. Ok.